File: sdk/java/src/test/java/world/navis/heartlight/SynapticKernelTest.java
package world.navis.heartlight;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class SynapticKernelTest {
    @Test
    void conformanceV1() {
        var state = new SynapticKernel.State(0.2, 0.0);
        var inputs = new SynapticKernel.Input[] {
            new SynapticKernel.Input(0.5, 0.25, 1.0),
            new SynapticKernel.Input(1.0, -0.5, 0.4),
            new SynapticKernel.Input(0.2, 0.8, -0.25)
        };
        double[] weights = {0.2096, 0.1967808, 0.2001622384};
        double[] traces = {0.125, -0.3875, -0.18875};
        for (int i = 0; i < inputs.length; i++) {
            state = SynapticKernel.step(state, inputs[i]);
            assertEquals(weights[i], state.weight(), 1e-12);
            assertEquals(traces[i], state.eligibility(), 1e-12);
        }
    }
}

File: sdk/kotlin/src/test/kotlin/heartlight/SynapticKernelTest.kt
package heartlight

import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertTrue

class SynapticKernelTest {
    @Test
    fun conformanceV1() {
        var state = SynapticState(0.2, 0.0)
        val inputs = listOf(
            SynapticInput(0.5, 0.25, 1.0),
            SynapticInput(1.0, -0.5, 0.4),
            SynapticInput(0.2, 0.8, -0.25),
        )
        val weights = listOf(0.2096, 0.1967808, 0.2001622384)
        val traces = listOf(0.125, -0.3875, -0.18875)
        inputs.forEachIndexed { index, input ->
            state = synapticStep(state, input)
            assertTrue(abs(state.weight - weights[index]) < 1e-12)
            assertTrue(abs(state.eligibility - traces[index]) < 1e-12)
        }
    }
}
